strawberryfields/backends/bosonicbackend/bosoniccircuit.py

```diff
@@ -0,0 +1,833 @@
+            self.means = np.delete(self.means, [2 * new_mode, 2 * new_mode + 1], axis=1)
+            self.nlen = self.nlen - 1
+            self.from_xp = from_xp(self.nlen)
+            self.to_xp = to_xp(self.nlen)
+            self.active = self.active[:new_mode]
+            prefac = -np.tan(theta) / np.sqrt(2 * self.hbar * eta_anc)
+            self.displace(prefac * val[0][0], np.pi / 2, k)
+
+        self.phase_shift(phi / 2, k)
+
+        if not avg:
+            return val
+
+        return None
+
+    def phase_shift(self, phi, k):
+        r"""Implement a phase shift in mode k.
+
+        Args:
+           phi (float): phase
+           k (int): mode to be phase shifted
+        """
+        if self.active[k] is None:
+            raise ValueError("Cannot phase shift mode, mode does not exist")
+
+        rot = symp.expand(symp.rotation(phi), k, self.nlen)
+        self.means = update_means(self.means, rot, self.from_xp)
+        self.covs = update_covs(self.covs, rot, self.from_xp)
+
+    def beamsplitter(self, theta, phi, k, l):
+        r"""Implement a beam splitter operation between modes k and l.
+
+        Args:
+            theta (float): real beamsplitter angle
+            phi (float): complex beamsplitter angle
+            k (int): first mode
+            l (int): second mode
+        """
+        if self.active[k] is None or self.active[l] is None:
+            raise ValueError("Cannot perform beamsplitter, mode(s) do not exist")
+
+        if k == l:
+            raise ValueError("Cannot use the same mode for beamsplitter inputs")
+
+        # Cross-check with Gaussian backend.
+        bs = symp.expand(symp.beam_splitter(theta, phi), [k, l], self.nlen)
+        self.means = update_means(self.means, bs, self.from_xp)
+        self.covs = update_covs(self.covs, bs, self.from_xp)
+
+    def scovmatxp(self):
+        r"""Returns the symmetric ordered array of covariance matrices
+        in the :math:`q_1,...,q_n,p_1,...,p_n` ordering.
+        """
+        return self.covs[:, self.to_xp][..., self.to_xp]
+
+    def smeanxp(self):
+        r"""Returns the symmetric ordered array of means in the
+        :math:`q_1,...,q_n,p_1,...,p_n` ordering.
+        """
+        return self.means.T[self.to_xp].T
+
+    def scovmat(self):
+        r"""Returns the symmetric ordered array of covariance matrices
+        in the :math:`q_1,p_1,...,q_n,p_n` ordering.
+        """
+        return self.covs
+
+    def smean(self):
+        r"""Returns the symmetric ordered array of means
+        in the :math:`q_1,p_1,...,q_n,p_n` ordering.
+        """
+        return self.means
+
+    def sweights(self):
+        """Returns the array of weights."""
+        return self.weights
+
+    def fromsmean(self, r, modes=None):
+        r"""Populates the array of means from a provided array of means.
+
+        The input must already have performed the scaling of the means by self.hbar,
+        and must be sorted in ascending order.
```

Comment:
Maybe worth being explicit here? Maybe mistaken, but ascending could be confused with the :math:`(x_1,\dots,x_n,p_1,\dots,p_n,)` ordering?
```suggestion
        and must be sorted :math:`(x_1,p_1,x_2,p_2,\dots)` order.
```

Reply:
Although the name is used in the Gaussian case, maybe worth considering `update_means`? The current name suggests that the method can only be used for initialization, however, it can serve as an update any time.